checker/checkInductive.ml

```diff
@@ -8,264 +8,154 @@
-   and that all arguments expected by this constructor are
-   logical, this is the case for equality, conjunction of logical properties
-*)
-let is_unit constrsinfos =
-  match constrsinfos with  (* One info = One constructor *)
-   | [|constrinfos|] -> Univ.is_type0m_univ constrinfos
-   | [||] -> (* type without constructors *) true
-   | _ -> false
-
-let small_unit constrsinfos =
-  let issmall = Array.for_all Univ.is_small_univ constrsinfos
-  and isunit = is_unit constrsinfos in
-  issmall, isunit
-
-let all_sorts = [InProp;InSet;InType]
-let small_sorts = [InProp;InSet]
-let logical_sorts = [InProp]
-
-let allowed_sorts issmall isunit s =
-  match Sorts.family s with
-  (* Type: all elimination allowed *)
-  | InType -> all_sorts
-
-  (* Small Set is predicative: all elimination allowed *)
-  | InSet when issmall -> all_sorts
+  let mind_entry_inds = Array.map_to_list (fun ind ->
+      let mind_entry_arity, mind_entry_template = match ind.mind_arity with
+        | RegularArity ar ->
+          let ctx, arity = Term.decompose_prod_n_assum nparams ar.mind_user_arity in
+          ignore ctx; (* we will check that the produced user_arity is equal to the input *)
+          arity, false
+        | TemplateArity ar ->
+          let ctx = ind.mind_arity_ctxt in
+          let ctx = List.firstn (List.length ctx - nparams) ctx in
+          Term.mkArity (ctx, Sorts.sort_of_univ ar.template_level), true
+      in
+      {
+        mind_entry_typename = ind.mind_typename;
+        mind_entry_arity;
+        mind_entry_template;
+        mind_entry_consnames = Array.to_list ind.mind_consnames;
+        mind_entry_lc = Array.map_to_list (fun c ->
+            let ctx, c = Term.decompose_prod_n_assum nparams c in
+            ignore ctx; (* we will check that the produced user_lc is equal to the input *)
+            c
+          ) ind.mind_user_lc;
+      })
+      mb.mind_packets
+  in
+  {
+    mind_entry_record;
+    mind_entry_finite = mb.mind_finite;
+    mind_entry_params = mb.mind_params_ctxt;
+    mind_entry_inds;
+    mind_entry_universes;
+    mind_entry_private = mb.mind_private;
+  }
+
+let check_arity ar1 ar2 = match ar1, ar2 with
+  | RegularArity ar, RegularArity {mind_user_arity;mind_sort} ->
+    Constr.equal ar.mind_user_arity mind_user_arity &&
+    Sorts.equal ar.mind_sort mind_sort
+  | TemplateArity ar, TemplateArity {template_param_levels;template_level} ->
+    List.equal (Option.equal Univ.Level.equal) ar.template_param_levels template_param_levels &&
+    Univ.Universe.equal ar.template_level template_level
+  | (RegularArity _ | TemplateArity _), _ -> false
+
+(* Use [eq_ind_chk] because when we rebuild the recargs we have lost
+   the knowledge of who is the canonical version.
+   Try with to see test-suite/coqchk/include.v *)
+let eq_recarg a1 a2 = match a1, a2 with
+  | Norec, Norec -> true
+  | Mrec i1, Mrec i2 -> eq_ind_chk i1 i2
+  | Imbr i1, Imbr i2 -> eq_ind_chk i1 i2
+  | (Norec | Mrec _ | Imbr _), _ -> false
```

Comment:
This is a bit adhoc but better than before (where we just ignored the recargs field).